src/hash_traits.jl

```diff
@@ -0,0 +1,461 @@
+# serialized but here we want that to happen by default, so e.g. ==(2) will properly hash
+# both the name of `==` and `2`.
+function transformer(::Type{<:Function}, ::HashVersion{3})
+    return Transformer(identity, StructTypes.UnorderedStruct())
+end
+
+function type_hash_name(::Type{T}, ::StructTypes.NoStructType) where {T<:Function}
+    return function_type_name(T)
+end
+function type_value_name(::Type{T}, ::StructTypes.NoStructType) where {T<:Function}
+    return function_type_name(T)
+end
+
+function function_type_name(::Type{T}) where {T}
+    if hasproperty(T, :instance) && isdefined(T, :instance)
+        return "typeof($(qualified_name_(T.instance)))"
+    else
+        return qualified_name_(T)
+    end
+end
+
+#####
+##### DataType
+#####
+
+sorted_field_names(T::Type) = TupleTools.sort(fieldnames(T); by=string)
+@generated function sorted_field_names(T)
+    return TupleTools.sort(fieldnames(T); by=string)
+end
+
+function type_structure(::Type{T}, trait::StructTypes.DataType) where {T}
+    if isconcretetype(T)
+        fields = trait isa StructTypes.OrderedStruct ? fieldnames(T) : sorted_field_names(T)
+        return fields, map(field -> fieldtype(T, field), fields)
+    else
+        return nothing
+    end
+end
+
+function stable_hash_helper(x, hash_state, context, st::StructTypes.DataType)
+    nested_hash_state = start_nested_hash!(hash_state)
+
+    # hash the field values
+    fields = st isa StructTypes.UnorderedStruct ? sorted_field_names(x) :
+             fieldnames(typeof(x))
+    nested_hash_state = hash_fields(x, fields, nested_hash_state, context)
+    hash_state = end_nested_hash!(hash_state, nested_hash_state)
+    return hash_state
+end
+
+function hash_fields(x, fields, hash_state, context)
+    for field in fields
+        val = getfield(x, field)
+        # can we optimize away the field's type_hash?
+        transform = transformer(typeof(val), context)
+        if isconcretetype(fieldtype(typeof(x), field)) && transform.preserves_structure
+            # the fieldtype has been hashed as part of the type of the container
+            hash_value(val, hash_state, context, transform)
+        else
+            hash_type_and_value(val, hash_state, context)
+        end
+    end
+    return hash_state
+end
+
+#####
+##### ArrayType
+#####
+
+is_ordered(x) = true
+is_ordered(::AbstractSet) = false
+order_by(x::Symbol) = String(x)
+order_by(x::Char) = string(x)
+order_by(x) = x
+
+function type_structure(::Type{T}, ::StructTypes.ArrayType) where {T}
+    return eltype(T)
+end
+
+# include ndims in type hash when we can
+function type_structure(::Type{T}, ::StructTypes.ArrayType) where {T<:AbstractArray}
+    return eltype(T), ndims_(T)
+end
+ndims_(::Type{<:AbstractArray{<:Any,N}}) where {N} = N
+ndims_(::Type{<:AbstractArray}) = nothing
+
+function transformer(::Type{<:AbstractRange}, ::HashVersion{3})
+    return Transformer(identity, StructTypes.Struct(); preserves_structure=true)
+end
```

Comment:
why do we need to special case this one? does it default to `UnorderedStruct`?

Reply:
(if we need to special case it that seems fine but we should document it)

Reply:
`UnorderedStruct == Struct`, but I will change the name for clarity. `AbstractRange` defaults to `ArrayType`, which leads to an inefficient hash. I've added docs to the README about this now.